internal/transports/p2p/peer/peer.go

```diff
@@ -0,0 +1,554 @@
+			if service.HeaderAlreadyExists.Is(err) {
+				continue
+			}
+
+			if service.BlockRejected.Is(err) {
+				// TODO: ban peer
+				p.log.Error().Msgf("received rejected header %v from peer %s", h, p)
+				_ = p.Disconnect()
+				return
+			}
+
+			if service.HeaderSaveFail.Is(err) {
+				p.log.Error().Msgf("couldn't save header %v in database, because of %+v", h, err)
+				continue
+			}
+
+			if service.HeaderCreationFail.Is(err) {
+				p.log.Error().Msgf("couldn't create header from %v because of error %+v", header, err)
+				continue
+			}
+
+			if service.ChainUpdateFail.Is(err) {
+				p.log.Error().Msgf("when adding header %v couldn't update chains state because of error %+v", header, err)
+				continue
+			}
+		}
+
+		receivedCheckpoint, err = p.verifyCheckpointReached(h, receivedCheckpoint)
+		if err != nil {
+			// TODO: ban peer or lower peer sync score
+			_ = p.Disconnect()
+			return
+		}
+
+		lastHeight = h.Height
+		headersReceived += 1
+	}
+
+	if headersReceived == 0 {
+		p.log.Debug().Msgf("received only existing headers from peer: %s", p)
+		return
+	}
+
+	p.log.Info().Msgf(
+		"successfully received %d headers from peer %s, up to height %d",
+		headersReceived, p, lastHeight,
+	)
+
+	p.requestNextHeadersBatch(receivedCheckpoint, lastHeight)
+}
+
+func (p *Peer) verifyCheckpointReached(h *domains.BlockHeader, receivedCheckpoint bool) (bool, error) {
+	if p.nextCheckpoint != nil && p.nextCheckpoint.Hash != nil && h.Height == p.nextCheckpoint.Height {
+		if h.Hash == *p.nextCheckpoint.Hash {
+			receivedCheckpoint = true
+			p.log.Info().Msgf(
+				"verified downloaded block header against checkpoint at height %d / hash %s",
+				h.Height, h.Hash,
+			)
+		} else {
+			p.log.Error().Msgf(
+				"block header at height %d/hash %s from peer %s does NOT match expected checkpoint hash of %s -- disconnecting",
+				h.Height, h.Hash, p, p.nextCheckpoint.Hash,
+			)
+			return false, fmt.Errorf("corresponding checkpoint height does not match got: %v, exp: %v", h.Height, p.nextCheckpoint.Height)
+		}
+	}
+	return receivedCheckpoint, nil
+}
+
+func (p *Peer) requestNextHeadersBatch(receivedCheckpoint bool, lastHeight int32) {
+	if receivedCheckpoint {
+		p.nextCheckpoint = findNextHeaderCheckpoint(p.checkpoints, p.nextCheckpoint.Height)
+		if p.nextCheckpoint == nil {
+			p.synced = true
+		}
+	}
+
+	if p.nextCheckpoint != nil {
+		p.log.Info().Msgf(
+			"requesting next headers batch from peer %s, height range %d - %d",
+			p, lastHeight, p.nextCheckpoint.Height,
+		)
+		err := p.writeGetHeadersMsg(p.nextCheckpoint.Hash)
+		if err != nil {
+			p.log.Error().Msgf("error requesting headers from peer %s, reason: %v", p, err)
+		}
+		return
+	}
+
+	p.log.Info().Msgf(
+		"checkpoints synced, requesting headers from height %d up to end of chain (zero hash) from peer %s",
+		lastHeight, p,
+	)
+
+	err := p.writeGetHeadersMsg(&zeroHash)
+	if err != nil {
+		p.log.Error().Msgf("error requesting headers from peer %s, reason: %v", p, err)
```

Comment:
We should add a subtask to handle properly errors when we cannot ask for another batch of headers,
because if we're not able to send `GetHeaders` message, then we won't receive anything, and the whole synchronisation will stuck, and we'ld be doomed 🙂 

But we can resolve that later in a separate task

Reply:
Okay, I added a TODO comment saying to lower the peer sync score.

BTW. everywhere I added the TODO comment saying "ban peer or lower peer sync score", I was having in mind that we will be disconnecting the peer and switching to another one for headers syncing. According to our idea about having a pool of open connections to peer, but only one (best) peer that we're syncing the headers from.

I will add a task for it, but we first need to implement the pool of connections in the server and we can't do both of these tasks in paralel.

Reply:
It's ok, 
Previously the error was only logged so, actually we didn't react to it correctly.